File: backend/DevTools.Domain/Entities/Command.cs
namespace DevTools.Domain.Entities;

public class Command
{
    public Guid Id { get; private set; }
    public string Title { get; private set; }
    public string CommandText { get; private set; }
    public string? Description { get; private set; }
    public string? Tags { get; private set; }
    public bool IsFavorite { get; private set; }
    public int UsageCount { get; private set; }
    public DateTimeOffset CreatedAt { get; private set; }
    public Guid CategoryId { get; private set; }
    public Guid UserId { get; private set; }

    private Command()
    {
        Title = string.Empty;
        CommandText = string.Empty;
    }

    public Command(string title, string commandText, string? description, string? tags, Guid categoryId, Guid userId)
    {
        Id = Guid.NewGuid();
        Title = title;
        CommandText = commandText;
        Description = description;
        Tags = tags;
        IsFavorite = false;
        UsageCount = 0;
        CreatedAt = DateTimeOffset.UtcNow;
        CategoryId = categoryId;
        UserId = userId;
    }

    public void Update(string title, string commandText, string? description, string? tags, Guid categoryId)
    {
        Title = title;
        CommandText = commandText;
        Description = description;
        Tags = tags;
        CategoryId = categoryId;
    }

    public void SetFavorite(bool isFavorite) => IsFavorite = isFavorite;

    public void IncrementUsage() => UsageCount++;
}

File: web/src/app/core/services/command.ts
import { HttpClient } from '@angular/common/http';
import { Service, inject, signal } from '@angular/core';
import { firstValueFrom } from 'rxjs';

import { API_BASE_URL } from '../config/api-config';
import type { Command, CommandRequest } from '../models/command.model';

@Service()
export class CommandService {
  private readonly http = inject(HttpClient);

  private readonly commandsSignal = signal<Command[]>([]);
  readonly commands = this.commandsSignal.asReadonly();

  async load(): Promise<void> {
    const commands = await firstValueFrom(this.http.get<Command[]>(`${API_BASE_URL}/commands`));
    this.commandsSignal.set(commands);
  }

  async create(request: CommandRequest): Promise<Command> {
    const command = await firstValueFrom(
      this.http.post<Command>(`${API_BASE_URL}/commands`, request),
    );
    this.commandsSignal.update((list) => [command, ...list]);
    return command;
  }

  async update(id: string, request: CommandRequest): Promise<Command> {
    const updated = await firstValueFrom(
      this.http.put<Command>(`${API_BASE_URL}/commands/${id}`, request),
    );
    this.commandsSignal.update((list) => list.map((c) => (c.id === id ? updated : c)));
    return updated;
  }

  async setFavorite(id: string, isFavorite: boolean): Promise<void> {
    const updated = await firstValueFrom(
      this.http.patch<Command>(`${API_BASE_URL}/commands/${id}/favorite`, { isFavorite }),
    );
    this.commandsSignal.update((list) => list.map((c) => (c.id === id ? updated : c)));
  }

  async incrementUsage(id: string): Promise<number> {
    const result = await firstValueFrom(
      this.http.post<{ id: string; usageCount: number }>(`${API_BASE_URL}/commands/${id}/use`, {}),
    );
    this.commandsSignal.update((list) =>
      list.map((c) => (c.id === id ? { ...c, usageCount: result.usageCount } : c)),
    );
    return result.usageCount;
  }

  async delete(id: string): Promise<void> {
    await firstValueFrom(this.http.delete<void>(`${API_BASE_URL}/commands/${id}`));
    this.commandsSignal.update((list) => list.filter((c) => c.id !== id));
  }

  /** Re-creates a just-deleted command, for the "undo" snackbar action. */
  async restore(command: Command): Promise<void> {
    await this.create({
      title: command.title,
      commandText: command.commandText,
      description: command.description,
      categoryId: command.categoryId,
      tags: command.tags,
    });
  }
}
